src/main/java/org/jenkinsci/plugins/scriptsecurity/scripts/ScriptApproval.java

```diff
@@ -668,6 +661,14 @@ public static void popRegistrationCallback() {
         callbacks.get().pop();
     }
 
+    @DataBoundSetter
+    public synchronized void setApprovedSignatures(String[] signatures) {
+        Jenkins.getInstance().checkPermission(Jenkins.RUN_SCRIPTS);
+        approvedSignatures.clear();
+        approvedSignatures.addAll(Arrays.asList(signatures));
+        save();
```

Comment:
I just realized that we never actually reconfigure `ApprovedWhitelist` based on these signatures until a sandboxed script runs for the first time. I think you should add a call to `reconfigure` here like the other methods that alter `approvedSignatures`.

Also, unlike the other methods that take signatures like `approveSignature`, in `setApprovedSignatures` users have direct control over the signature itself, so it might also be a good idea to do something like this to filter out any totally malformed signatures:

```
List<String> goodSignatures = new ArrayList<>(signatures.length);
for (String signature: signatures) {
  try {
    StaticWhitelist.parse(signature); // Method visibility might need to be changed to package-private
    goodSignatures.add(signature);
  } catch (IOException e) {
    LOGGER.log(Level.WARNING, "Ignoring malformed signature: " + signature);
  }
  ... the rest of the method ...
}
```

If you don't do this (but you do add the call to `reconfigure` after save), and a malformed signature does end up being saved, then approval of signatures from the UI will be broken until the malformed signature is removed because an exception will be thrown when `reconfigure` is called from `approveSignature` or similar.

Reply:
Good idea, I'll fix that...